audit_test.go

```diff
@@ -864,14 +891,115 @@ func TestRuleParsing(t *testing.T) {
 		if err != nil {
 			t.Fatal(err, msg)
 		}
-		cmdline, err := rule.ToCommandLine(data, true)
+		addRule := true
+		switch r.TypeOf() {
+		case rule.DeleteFileWatchRuleType, rule.DeleteSyscallRuleType:
+			addRule = false
+		}
+		cmdline, err := rule.ToCommandLineAddRemove(data, true, addRule)
 		if err != nil {
 			t.Fatal(err, msg)
 		}
 		assert.Equal(t, line, cmdline, msg)
 	}
 }
 
+func TestInvalidRuleParsing(t *testing.T) {
+	rules := []string{
+		"-D -a",
+		"-D -A",
+		"-D -d",
+
+		"-D -a always",
+		"-D -A always",
+		"-D -d always",
+
+		"-D -a never",
+		"-D -A never",
+		"-D -d never",
+
+		"-D -a always,task",
+		"-D -A always,task",
+		"-D -d always,task",
+
+		"-D -a always,task -w /foo/bar -p rw",
+		"-D -a always,task -W /foo/bar -p rw",
+		"-D -A always,task -w /foo/bar -p rw",
+		"-D -A always,task -W /foo/bar -p rw",
+		"-D -d always,task -w /foo/bar -p rw",
+		"-D -d always,task -W /foo/bar -p rw",
+
+		"-D -a never,task",
+		"-D -A never,task",
+		"-D -d never,task",
+
+		"-D -w /foo/bar",
+		"-D -W /foo/bar",
+		"-D -w /foo/bar -p rw",
+		"-D -W /foo/bar -p rw",
+		"-D -w /foo/bar -p garbage",
+		"-D -W /foo/bar -p garbage",
+
+		"-w /foo/bar -W /foo/bar",
+		"-w /foo/bar -W /foo/bar -p rw",
+
+		"-w foo/bar",
+		"-W foo/bar",
+		"-w foo/bar -p rw",
+		"-W foo/bar -p rw",
+
+		"-w foo/bar -S 42",
+		"-W foo/bar -S 42",
+
+		"-w foo/bar -W foo/bar",
+		"-w foo/bar -W foo/bar -p rw",
+
+		"-w foo/bar -W foo/bar -S 42",
+
+		"-w /foo/bar -F uid=100",
+		"-W /foo/bar -F uid=100",
+
+		"-w /foo/bar -S 42",
+		"-W /foo/bar -S 42",
+
+		"-w /foo/bar -F uid=100",
+		"-W /foo/bar -F uid=100",
+
+		"-w /foo/bar -C auid!=uid",
+		"-W /foo/bar -C auid!=uid",
+
+		"-a always,exit -w /foo/bar -p rw",
+		"-a always,exit -W /foo/bar -p rw",
+		"-A always,exit -w /foo/bar -p rw",
+		"-A always,exit -W /foo/bar -p rw",
+
+		"-a always,exit -w /foo/bar",
+		"-a always,exit -W /foo/bar",
+		"-A always,exit -w /foo/bar",
+		"-A always,exit -W /foo/bar",
+
+		"-a filesystem,always", // actually valid but there's no support for it
+		"-a filesystem,never",  // actually valid but there's no support for it
+	}
+
+	t.Logf("checking %d rules", len(rules))
+	for idx, line := range rules {
+		r, err := flags.Parse(line)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		_, err = rule.Build(r)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+
+		t.Fatalf("parsing line #%d: `%s' should have failed", idx, line)
```

Comment:
```suggestion
		t.Errorf("parsing line #%d: `%s' should have failed", idx, line)
```
https://github.com/golang/go/wiki/TestComments#keep-going

Reply:
Done